### PR TITLE
Enhanced authentication

### PR DESCRIPTION
**Pull request by tigercl**

### CODE REVIEW

File: test/emqx_auth_username_SUITE.erl [hunk 1]
```diff
@@ -42,11 +48,243 @@ end_per_suite(_Config) ->
...
+                                        client_id   = <<"mqtt_client">>,
+                                        proto_ver   = ?MQTT_PROTO_V5,
+                                        clean_start = false,
+                                        properties  = #{'Authentication-Method' => <<>>,
+                                                        'Authentication-Data'   => <<>>}})),
+    emqx_client_sock:send(Sock1, Packet1),
+    {ok, Data1} = gen_tcp:recv(Sock1, 0),
+    {ok, ?CONNACK_PACKET(?RC_MALFORMED_PACKET), _} = raw_recv_pase(Data1),
+    emqx_client_sock:close(Sock1),
+
```
Comment (Gilbert-Wong) [inline test/emqx_auth_username_SUITE.erl]: There are too many duplicated procedure in your test cases, can you abstract your code and make it shorter and more clear ?

Comment (Gilbert-Wong) [reply test/emqx_auth_username_SUITE.erl]: You can just write like this
```erlang
with_connection(fun(Sock) ->

                   %% your procedure
                end).

```

Comment (tigercl) [reply test/emqx_auth_username_SUITE.erl]: It's difficult, because it's flexible.

File: src/emqx_auth_username.erl [hunk 2]
```diff
@@ -114,6 +114,32 @@ init(Userlist) ->
...
+    {error, invalid_auth_data};
+check(#{username := Username, auth_method := <<"PLAIN">>, auth_data := AuthData}, _Password, _Opts) ->
+    case binary:split(AuthData, <<0>>, [global, trim]) of
+        [_, Username1, Password] -> 
+            Username2 = case Username of 
+                            undefined ->
+                                Username1;
+                            _->
+                                Username
+                        end,
```
Comment (terry-xiaoyu) [inline src/emqx_auth_username.erl]: And useful trick to do this is using function.

**Pull request closed without merging**